File: bot/core/logs/monitoring/advancements.py
import discord
from deep_translator import GoogleTranslator

async def process_advancements_messages(log_line, channel):
    try:

        ignore_patterns = [
            "[Rcon]", "[Not Secure]", "Disconnecting VANILLA connection attempt",
            "rejected vanilla connections", "lost connection", "id=<null>", "legacy=false",
            "lost connection: Disconnected", "<init>", "<", ">"
        ]

        if any(pattern in log_line for pattern in ignore_patterns):
            return

        player_name, message, event_name = extract_player_message(log_line)

        if player_name and message and event_name:

            if "has reached the goal" in message:
                await event_translation(channel, player_name, event_name, 0xFFA500, "alcançou um objetivo:")
            elif "has made the advancement" in message:
                await event_translation(channel, player_name, event_name, 0x0000FF, "obteve um avanço:")
            elif "has completed the challenge" in message:
                await event_translation(channel, player_name, event_name, 0x800080, "completou um desafio:")

        return
    except Exception as e:
        print(f"Erro ao processar evento de usuários mandando mensagens no discord: {e}")

def extract_player_message(log_line):
    try:
        message_part = log_line.split("]: ")[1]

        player_name_end = message_part.index(" has ")
        player_name = message_part[:player_name_end]

        message = message_part[player_name_end:].split("[")[0].strip()

        event_name_start = message_part.index("[") + 1
        event_name_end = message_part.index("]")
        event_name = message_part[event_name_start:event_name_end].strip()

        return player_name, message, event_name
    except ValueError:
        return None, None, None


async def event_translation(channel, player_name, event_name, color, action):
    event_translated = GoogleTranslator(source='en', target='pt').translate(event_name)
    await send_player_event(channel, player_name, action, event_name, event_translated, color)


async def send_player_event(channel, player_name, event_message, event_name, event_translated, color):
    try:
        embed = discord.Embed(color=color)
        embed.set_author(
            name=f'{player_name} {event_message} {event_name} ({event_translated})',
            icon_url=f"https://mineskin.eu/helm/{player_name}"
        )
        await channel.send(embed=embed)
    except Exception as e:
        print(f"[BOT ERROR] Falha ao enviar evento do jogador como embed: {e}")

File: internal/presentation/advancements.go
package presentation

import (
	"discord-bot-go/internal/connections"
	"discord-bot-go/internal/log"
	"fmt"
	"strings"

	"github.com/bwmarrin/discordgo"
)

func ProcessAdvancementLine(s *discordgo.Session, channelID string, line string) {
	ignorePatterns := []string{
		"[Rcon]", "[Not Secure]", "Disconnecting VANILLA connection attempt",
		"rejected vanilla connections", "lost connection", "id=<null>", "legacy=false",
		"lost connection: Disconnected", "<init>", "<", ">",
	}

	for _, pat := range ignorePatterns {
		if strings.Contains(line, pat) {
			return
		}
	}

	playerName, message, eventName := extractPlayerMessage(line)
	if playerName == "" || message == "" || eventName == "" {
		return
	}

	var color int
	var action string

	if strings.Contains(message, "has reached the goal") {
		color = 0xFFA500
		action = "alcançou um objetivo:"
	} else if strings.Contains(message, "has made the advancement") {
		color = 0x0000FF
		action = "obteve um avanço:"
	} else if strings.Contains(message, "has completed the challenge") {
		color = 0x800080
		action = "completou um desafio:"
	} else {
		return
	}

	sendPlayerEvent(s, channelID, playerName, action, eventName, color)
}

func extractPlayerMessage(logLine string) (string, string, string) {
	parts := strings.SplitN(logLine, "]: ", 2)
	if len(parts) < 2 {
		return "", "", ""
	}

	msgPart := parts[1]

	playerEnd := strings.Index(msgPart, " has ")
	if playerEnd == -1 {
		return "", "", ""
	}
	playerName := msgPart[:playerEnd]

	message := msgPart[playerEnd:]
	eventStart := strings.Index(msgPart, "[")
	eventEnd := strings.Index(msgPart, "]")
	if eventStart == -1 || eventEnd == -1 || eventEnd <= eventStart {
		return playerName, message, ""
	}
	eventName := msgPart[eventStart+1 : eventEnd]

	return playerName, message, eventName
}

func sendPlayerEvent(
	s *discordgo.Session,
	channelID string,
	playerName string,
	action string,
	eventName string,
	color int,
) {
	reader, filename, err := connections.GetPlayerImage(playerName)
	if err != nil {
		log.LogError("Falha ao obter imagem do player:", err)
		return
	}

	file := &discordgo.File{
		Name:   filename,
		Reader: reader,
	}

	embed := &discordgo.MessageEmbed{
		Color: color,
		Author: &discordgo.MessageEmbedAuthor{
			Name:    fmt.Sprintf("%s %s %s", playerName, action, eventName),
			IconURL: "attachment://" + filename,
		},
	}

	_, err = s.ChannelMessageSendComplex(
		channelID,
		&discordgo.MessageSend{
			Embed: embed,
			Files: []*discordgo.File{file},
		},
	)

	if err != nil {
		log.LogError("Falha ao enviar evento do jogador:", err)
		return
	}
}
